core/src/main/scala/org/apache/spark/serializer/KryoSerializer.scala

```diff
@@ -324,6 +326,36 @@ class KryoDeserializationStream(
       }
     }
   }
+
+  final override def asIterator: Iterator[Any] = new NextIterator[Any] {
+    override protected def getNext() = {
+      if (KryoDeserializationStream.this.hasNext) {
+        readValue[Any]()
+      } else {
+        finished = true
+        null
+      }
+    }
+
+    override protected def close(): Unit = {
+      KryoDeserializationStream.this.close()
+    }
+  }
+
+  final override def asKeyValueIterator: Iterator[(Any, Any)] = new NextIterator[(Any, Any)] {
+    override protected def getNext() = {
+      if (KryoDeserializationStream.this.hasNext) {
+        (readKey[Any](), readValue[Any]())
```

Comment:
Given we are fix this, not make assumptions that if key is present, value will be as well ?

Reply:
You mean that if only a key exist we just ignore it like the current implementation would?

Reply:
Or potentially do something better.

```
  if (hasNext) {
    val key = readKey()
    if (hasNext) {
      return (key, readValue())
    }
  }
```